replicator/run.py

```diff
@@ -209,29 +218,33 @@ def perform_full_zone_sync(self, ns: Nameserver):
             if local_serial is None or local_serial < self.serials.get(zone, 0)
         }
 
-        # Apply additions
-        for zone in additions:
-            print(f'Adding zone {zone} ...')
-            ns.add_zone(zone)
-            local_zones.add(zone)
-
-        # Apply modifications
-        for zone in modifications:
-            print(f'Queueing AXFR for stale zone {zone} ...')
-            ns.axfr(zone)
-
-        # Apply deletions
-        for zone in deletions:
-            print(f'Deleting zone {zone} ...')
-            ns.remove_zone(zone)
-            local_zones.discard(zone)
-            print(f'Zone {zone} deleted.')
+        if additions or deletions:
+            with ns_class() as ns:
+                # Apply additions
+                for zone in additions:
+                    print(f'Adding zone {zone} ...')
+                    ns.add_zone(zone)
+                    local_zones.add(zone)
+
+                # Apply deletions
+                for zone in deletions:
+                    print(f'Deleting zone {zone} ...')
+                    ns.remove_zone(zone)
+                    local_zones.discard(zone)
+                    print(f'Zone {zone} deleted.')
+
+        if modifications:
+            with ns_class() as ns:
```

Comment:
why still use two instances of `ns_class()` instead of one?

Reply:
Because zone addition/removal needs a Knot config transaction (which may fail), whereas zone transfer happens outside of such a transaction and I wanted to make sure that zone transfers continue to work even when a zone can't be added/deleted for some reason.

I'm not sure if an aborted config transaction would really interfere with an AXFR command on the same socket connection, but I was too lazy to test.

There's also a slightly lower chance of running into the socket timeout this way (which shouldn't happen normally anyway).